### PR TITLE
Lay groundwork for POJO dari impl

### PR DESCRIPTION
**Pull request by efronbs**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Issue: https://github.com/yahoo/fili/issues/769

### CODE REVIEW

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/rfc/Generator.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * to generate. In your {@link AbstractBinderFactory} implementation you must make a named binding for the type of
+ * resource you wish to generate. You can only bind one generator per resource (though different resources may share
+ * a type, such as "count" and "topN"), so your implementation will replace the default generator for that resource.
+ *
+ * TODO: binding names should be written here for reference.
+ * TODO: binding order should be written here
+ *
+ * @param <T> The type of the resource this generator builds.
+ */
+public interface Generator<T> {
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/rfc/Generator.java]: This interface probably deserves a slightly more specific name.  I'm not thinking of one yet, though, so this is unhelpful feedback at the moment.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/rfc/RequestParameters.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.ws.rs.core.PathSegment;
+
+/**
+ * Data object all raw pieces of data in an api request.
+ */
+public class RequestParameters {
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/rfc/RequestParameters.java]: it's generally considered more appropriate to hold the nulls internally and optional wrap on return.  I don't remember if we discussed making an exception for this class or not.

Comment (michael-mclawhorn) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/rfc/RequestParameters.java]: I suppose given the write to read ratio on this object, it actually saves cycles to wrap once.

Comment (efronbs) [reply fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/rfc/RequestParameters.java]: yeah that was the motivation. It's an immutable object that has no logic behind its internal state. Storing optionals over nulls just saves a little memory on read.

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DataApiRequestValueObject.java [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.ws.rs.core.Response;
+
+/**
+ * And immutable POJO implementation of {@link DataApiRequest} contract. All data is provided through the constructor
+ * and set, with very minor or no transformations occurring on the data. Unlike the old implementation
+ * ({@link com.yahoo.bard.webservice.web.apirequest.DataApiRequestImpl}), this implementation does not build any of
+ * tis components and all components must be fully built at creation time.
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DataApiRequestValueObject.java]: spelling 'its'

File: fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DataApiRequestValueObject.java [hunk 4]
```diff
@@ -0,0 +1,772 @@
...
+    private final Granularity granularity;
+    private final LinkedHashSet<Dimension> dimensions;
+    private final LinkedHashMap<Dimension, LinkedHashSet<DimensionField>> perDimensionFields;
+    private final LinkedHashSet<LogicalMetric> metrics;
+    private final List<Interval> intervals;
+    private final ApiFilters apiFilters;
+    private final LinkedHashMap<LogicalMetric, Set<ApiHaving>> havings;
+    private final OrderByColumn dateTimeSort;
+    private final LinkedHashSet<OrderByColumn> standardSorts;
+    private final LinkedHashSet<OrderByColumn> allSorts;
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DataApiRequestValueObject.java]: Why are we keeping all three of these?  We should either keep the separate fields or the merged one.

File: CHANGELOG.md [hunk 5]
```diff
@@ -421,6 +421,10 @@ select filters)
 
 ### Added:
 
+- [Added groundwork classes for POJO `DataApiRequest` build path](https://github.com/yahoo/fili/issues/769)
+    * Generator contract, builder, and POJO object have all been added, along with some silent contract changes.
+    * Immutable implementations of `LinkedHashSet`, `LinkedHashMap`, and `ApiFilters` have been added. 
+
```
Comment (michael-mclawhorn) [inline CHANGELOG.md]: Need some additional CHANGELOG:

I would call out in the changed section:

Deprecated Optionals on constructors and withers
Replaced OptionalInt with Optional<Integer> on some interfaces

File: fili-core/src/main/java/com/yahoo/bard/webservice/config/BardFeatureFlag.java [hunk 6]
```diff
@@ -34,7 +34,8 @@
     DRUID_DIMENSIONS_LOADER("druid_dimensions_loader_enabled"),
     CASE_SENSITIVE_KEYS("case_sensitive_keys_enabled"),
     DEFAULT_IN_FILTER("default_in_filter_enabled"),
-    REQUIRE_METRICS_QUERY("require_metrics_in_query")
+    REQUIRE_METRICS_QUERY("require_metrics_in_query"),
+    POJO_DARI_REQUIRE_ALL_STAGES_CALLED("require_all_data_request_building_stages_called")
```
Comment (michael-mclawhorn) [inline fili-core/src/main/java/com/yahoo/bard/webservice/config/BardFeatureFlag.java]: We should add this to the default moduleConfig.properties file.

I'm inclined to default it to true, but I don't insist.

**Pull request merged**